File: 0895-maximum-frequency-stack/0895-maximum-frequency-stack.java
class FreqStack {
        private HashMap<Integer, Integer> freqMap;
        private int maxFreq;
        private HashMap<Integer, ArrayList<Integer>> elementMap;

    public FreqStack() {
        freqMap = new HashMap<>();
        maxFreq = 0;
        elementMap = new HashMap<>();
    }
    
    public void push(int val) {
        if(!freqMap.containsKey(val)){
            freqMap.put(val, 1);
        }else{
            int freq = freqMap.get(val);
            freqMap.put(val, freq+1);
        }

        maxFreq = Math.max(maxFreq, freqMap.get(val));

        if(!elementMap.containsKey(freqMap.get(val))){
            elementMap.put(freqMap.get(val), new ArrayList<>());
            elementMap.get(freqMap.get(val)).add(val);
        }else{
            elementMap.get(freqMap.get(val)).add(val);
        }
    }
    
    public int pop() {
        ArrayList<Integer> arr = elementMap.get(maxFreq);
        int result = arr.get(arr.size() - 1);
        arr.remove(arr.size() - 1);
        freqMap.put(result, freqMap.get(result) -1);

        if(arr.isEmpty()){
            elementMap.remove(maxFreq);
        
            maxFreq--;
        }
        return result;
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(val);
 * int param_2 = obj.pop();
 */

File: 0895-maximum-frequency-stack/0895-maximum-frequency-stack.js

var FreqStack = function() {
    this.freqMap = {};
    this.maxFrequency = 0;
    this.elementMap = {};
};

/** 
 * @param {number} val
 * @return {void}
 */
FreqStack.prototype.push = function(val) {
    if(!this.freqMap[val]){
        this.freqMap[val] = 1;
        
    }else{
        this.freqMap[val] += 1;
    }

    this.maxFrequency = Math.max(this.maxFrequency,  this.freqMap[val] );
    const frequency = this.freqMap[val];
    if(!this.elementMap[frequency]){
        this.elementMap[frequency] = [val];
    }else{
        this.elementMap[frequency].push(val);
    }
};

/**
 * @return {number}
 */
FreqStack.prototype.pop = function() {
    const arr = this.elementMap[this.maxFrequency]
    let result = arr[arr.length -1];
    this.elementMap[this.maxFrequency].pop();  /*i can also directly pop it and store it in result 
                                                but some stack in other lang. does not support to return that element 
                                                when pop function is run */
    this.freqMap[result]--;
    
    if(this.elementMap[this.maxFrequency].length == 0){
        delete this.elementMap[this.maxFrequency];

        this.maxFrequency -= 1;
    }

    return result;
};

/** 
 * Your FreqStack object will be instantiated and called as such:
 * var obj = new FreqStack()
 * obj.push(val)
 * var param_2 = obj.pop()
 */